### ISSUE TITLE
[stage/04] Training — Unsloth QLoRA fine-tuning to GGUF

### ISSUE THREAD

**Issue by Kabuswe** [Kabuswe/pipeline-finetune#5]
# [stage/04] Training — Unsloth QLoRA fine-tuning to GGUF

## Purpose
Implement `stages/s04_train.py`. Loads the JSONL dataset from Stage 03, fine-tunes the base model using Unsloth QLoRA, exports to GGUF with the configured quantization, and writes a ready-to-use `Modelfile`. This stage requires a CUDA GPU — designed to run in a Colab notebook, a cloud GPU instance, or a local machine with a capable GPU.

## File Structure
```
stages/
  s04_train.py              ← stage runner
trainer/
  __init__.py
  setup.py                  ← Unsloth model + tokenizer loading
  dataset_loader.py         ← JSONL → HuggingFace Dataset
  train_loop.py             ← SFTTrainer configuration and execution
  exporter.py               ← GGUF export + Modelfile generation
notebooks/
  colab_train.ipynb         ← Colab-ready notebook wrapping the stage
```

## Setup (`trainer/setup.py`)
```python
from unsloth import FastLanguageModel

def load_model(config: TrainingConfig):
    model, tokenizer = FastLanguageModel.from_pretrained(
        model_name=config.baseModel,   # e.g. "unsloth/llama-3.2-3b-instruct-bnb-4bit"
        max_seq_length=config.maxSeqLength,
        load_in_4bit=True,
    )
    model = FastLanguageModel.get_peft_model(
        model,
        r=config.loraR,
        lora_alpha=config.loraAlpha,
        target_modules=["q_proj", "k_proj", "v_proj", "o_proj",
                        "gate_proj", "up_proj", "down_proj"],
        use_gradient_checkpointing="unsloth",
    )
    return model, tokenizer
```

## Dataset Loader (`trainer/dataset_loader.py`)
- Load `datasets/{domain}_training.jsonl`
- Apply the model's chat template to convert `messages` arrays to formatted strings
- Shuffle with fixed seed for reproducibility
- Split: 95% train / 5% validation (separate from held-out eval set)

## Training Loop (`trainer/train_loop.py`)
```python
from trl import SFTTrainer
from transformers import TrainingArguments

trainer = SFTTrainer(
    model=model,
    tokenizer=tokenizer,
    train_dataset=train_dataset,
    eval_dataset=val_dataset,
    dataset_text_field="text",
    max_seq_length=config.maxSeqLength,
    args=TrainingArguments(
        per_device_train_batch_size=config.batchSize,
        gradient_accumulation_steps=config.gradientAccumulationSteps,
        max_steps=config.maxSteps,
        learning_rate=config.learningRate,
        fp16=not is_bfloat16_supported(),
        bf16=is_bfloat16_supported(),
        logging_steps=10,
        eval_strategy="steps",
        eval_steps=50,
        save_strategy="steps",
        save_steps=100,
        output_dir="output/checkpoints",
        report_to="none",           # no wandb dependency
    ),
)
```

## Export (`trainer/exporter.py`)
```python
def export_gguf(model, tokenizer, config: PipelineConfig):
    output_path = f"output/{config.outputModelName}-{config.quantization}"
    model.save_pretrained_gguf(
        output_path,
        tokenizer,
        quantization_method=config.quantization   # q4_k_m | q8_0 | q5_k_m
    )
    # Generates: output/{name}.gguf + Modelfile (auto by Unsloth)
    patch_modelfile(output_path + "/Modelfile", config.systemPrompt)
    return output_path + ".gguf"

def patch_modelfile(modelfile_path: str, system_prompt: str):
    """Inject the configured system prompt and inference parameters."""
    with open(modelfile_path, 'a') as f:
        f.write(f'\nSYSTEM """{system_prompt}"""\n')
        f.write('\nPARAMETER temperature 0.1\n')
        f.write('PARAMETER num_ctx 8192\n')
```

## Registry Write
After successful export, write a `ModelRecord` to the registry:
```python
upsert_model(ModelRecord(
    model_id=sha256(config.domain + utcnow()),
    domain=config.domain,
    base_model=config.training.baseModel,
    gguf_path=gguf_path,
    ollama_name=None,     # set in Stage 06 after ollama create
    accuracy=None,        # set in Stage 05
    ece_score=None,
    abstention_recall=None,
    promoted=False,
    created_at=utcnow()
))
```

## Colab Notebook (`notebooks/colab_train.ipynb`)
A self-contained notebook that:
1. Clones the repo and installs dependencies
2. Mounts Google Drive to read `datasets/` and write `output/`
3. Calls `run.py --from-stage 4 --to-stage 4`
4. Downloads the GGUF from Drive to local machine

## GPU Requirements by Model Size
| Base Model | VRAM (QLoRA 4-bit) | Quantized GGUF Size |
|---|---|---|
| 3B | ~4 GB | ~2 GB |
| 7B | ~8 GB | ~4.5 GB |
| 13B | ~14 GB | ~8 GB |
| 32B | ~22 GB | ~20 GB |

These map directly to the hardware recommendation service offering.

## Acceptance Criteria
- Fine-tuning completes on a T4 (Colab free tier) for the 3B model with default config
- GGUF file is present at `output/{outputModelName}-{quantization}.gguf`
- `Modelfile` contains the configured `systemPrompt`
- `ModelRecord` is written to registry with `gguf_path` populated
- Training loss is logged every 10 steps to `output/training_log.json`
- `--from-stage 4 --to-stage 4` skips training if a model for this domain was trained within the last 24 hours (staleness check via registry)

## Dependencies
- Issue #4 (Stage 03 JSONL) — `datasets/{domain}_training.jsonl` must exist

## Labels
`stage/04-training` `priority/p1`
